### ISSUE TITLE
Run a current version of Ansible in the Admin Workstation

### ISSUE THREAD

**Issue by conorsch** [freedomofpress/securedrop#1146]
At time of writing, Tails 1.6 ships with Ansible 1.7.2 available in the Debian stable repository. Feature-wise, 1.7.2 is recent enough for our playbooks—the apache2_module was added in 1.6, for example—but it lacks several important bugfixes.

One example is [Get MemoryError when fetching large file](https://github.com/ansible/ansible/issues/11702), which was patched as of 1.7.3. I've seen this bug cause failures with the automatic backup role that runs at the end of the playbook. The failure aborts the backup process, and no encrypted tarball is written to the Admin Workstation. In order to force the backup, a manual rsync command is required, which sidesteps the MemoryError exception.

The problem is how to install a recent version of Ansible safely under Debian wheezy (which Tails 1.6 is based on). The [Ansible install docs](http://docs.ansible.com/ansible/intro_installation.html) include a long list of installation methods, but none includes GPG verification, which the current apt-get install process provides.

**Comment by garrettr** [comment]
Potential solution worth investigating: as mentioned in https://github.com/freedomofpress/securedrop/issues/1035#issuecomment-140172267, Tails includes the Debian testing repos in their apt sources list by default. The current version of Ansible in Debian testing is [1.9.4](https://packages.debian.org/stretch/ansible).

**Comment by conorsch** [comment]
Yes, `sudo apt-get install -y -s testing ansible` works! Haven't tested apt persistence in Tails with more recent version, will try that next.

**Comment by garrettr** [comment]
@conorsch I bet the apt persistence would work just fine for packages installed from the testing repos. Have you had a chance to test it yet?

**Comment by garrettr** [comment]
You should also test the workflow for migrating Admin Workstations that were using ansible from the stable repos, since that package will be cached locally due to the apt persistence. Can they just run `apt-get install -t testing ansible`, or do they need to explicitly remove the stable ansible package first?

**Comment by conorsch** [comment]
@garrettr Haven't tested sufficiently yet. Persistence seems to work fine, but as you mention, it may be the stable version getting installed—haven't confirmed that it's the 1.9.x release that's getting reinstalled.

Also, installing ansible from the testing repo results in several interactive dpkg prompts. Sticking to the default suggestions for the confirmation prompts works fine, but will those prompts cause problems with apt persistence? Hopefully the Tails apt persistence correctly uses a Debian noninteractive environment, but will need to confirm to be sure.

**Comment by fowlslegs** [comment]
I installed testing/ansible on our test instance Admin Workstation. After reboot it reinstalled oldstable/ansible.

**Comment by heartsucker** [comment]
Commenting here since I just got bit by this while testing something. If we update, can we at least push over to something from the 2.1.x line? This is mostly backwards compatible with 1.8.4 (current recommendation), and will at at least get us the new goodies and syntax. This can be installed from `pip` easily (but not `apt`) for Debian 8 based systems.

**Comment by conorsch** [comment]
Need to loop back on this. We're still expecting a very old version of Ansible, and if persistence works with a 2.x release of Ansible, that's ideal. If not, I'd be willing to fall back to a `pip install` with an explicit version requirement. We'd lose GPG verification via apt, but with Tails forcing all connections over Tor, and pip now supporting TLS connections, we can reasonably assume it would be prohibitively difficult to perform a targeted MITM attack on the ansible install.

**Comment by fowlslegs** [comment]
Does Tails come w/ `python-pip` installed? If not is the plan to install that first and then use it to install Ansible? In Debian Jessie `python-pip` is currently version 1.5.6. I think that verison use HTTPS by default for PyPi, but it wouldn't hurt to confirm.

**Comment by fowlslegs** [comment]
There's an [official Ansible PPA](https://launchpad.net/~ansible/+archive/ubuntu/ansible) that would allow us to maintain GPG verification, while providing the latest stable Ansible release.

**Comment by conorsch** [comment]
> There's an official Ansible PPA that would allow us to maintain GPG verification, while providing the latest stable Ansible release.

@fowlslegs I like that idea, mainly because it's technically more secure than pip installations, but we'd still need to bootstrap the configuration of that repo _without_ Ansible to get it installed and configured. Given rickety state of the `install.sh` bash script, those changes should be very well tested.

**Comment by fowlslegs** [comment]
We talked and decided that with the precarious state of Tails APT sources and preferences lists, we are weary of further complicating this.

**Comment by conorsch** [comment]
@fowlslegs If the `become` -> `sudo` changes are removed, the `with_items` updates are still worth considering for inclusion. They _should_ be backwards compatible with all relevant prior versions of Ansible, but will need to be explicitly tested and confirmed working with:

* ansible==1.8.4 in staging VMs
* ansible==1.7.2 (via apt) on Tails

If the changes check out in both those contexts, let's get them in.

**Comment by conorsch** [comment]
First-time Admins frequently install Ansible v2.x via pip and then report problems, so updating the playbooks and just using the new-style syntax seems the best way to go. We can provide a fail task during validation to bail out of an old version of Ansible is used, so we don't have to support syntaxes for both versions.

**Comment by fowlslegs** [comment]
^ Seems like the way to go. Do you think we should persistently install Ansible on the AW, or just have admins install it ad-hoc when they need to run playbooks?

**Comment by conorsch** [comment]
We'd have to:

1. Ensure Ansible via apt is _not_ installed
2. Ensure the ansible package is absent from apt persistence settings
3. Install python-pip via apt
4. Install Ansible via python-pip

Theoretically that's all possible in a run of the `./install.sh` script, but it will naturally require explicit docs. This has me thinking more about the underlying problem, which is our shoehorning persistent system configs into an amnesiac OS....

**Comment by conorsch** [comment]
In #1531, @msheiny recommended bootstrapping a virtualenv to store the pip-version of Ansible in Tails's persistence:

> I imagine the below actions will be kicked off by a `./setup.sh` and user wont have to bother with the logic involved:
>* Install virtualenv on tails (its fine if this is blown away after reboot)
>* build a virtualenv environment under the persistent space, pip install ansible and any other third-party libraries we may need there
>* layout wrapper scripts that users call to kick off playbook launch for all runs that make use of this virtualenv. 

That sounds like a great idea. We'd need to require Admin intervention during the 0.4 release, but we've already been expecting that anyway. (Most 0.3.x releases have been strictly via auto updates.) We should make sure the playbooks detect and fail out if an unsuitable version of Ansible is detected.

**Comment by fowlslegs** [comment]
One thing to keep in mind in such a script is that it'll need to have logic which identifies if the right version of Ansible is installed to the virtualenv, and if not 
1. installs `python-pip` and `python-virtualenv` with apt,
2. updates `ansible` in the existing virtualenv, and
3. continues as normal.

**Comment by conorsch** [comment]
Tails 3 will be based on Debian 9, which includes Ansible 2.2.1 in the apt repos. We should _consider_ using that version to simplify the Admin Workstation setup—although even 2.2.1 will quickly become out of date, given the longevity of Debian releases.

@redshiftzero @msheiny Let's use 2.2.1 locally in the near-term to suss out problems.

**Comment by msheiny** [comment]
Want me to open a PR just to peg and test `2.2.1` in CI or should I append it to PR #1723 ?

**Comment by conorsch** [comment]
Peg 2.2.1 in #1723 and see if it passes. Seems like existing peg of 2.2.2 breaks testinfra locally—although not in CI.

**Comment by ageis** [comment]
@conorsch @redshiftzero Hey guys, is this rsync issue supposed to be fixed in develop branch with the trusty bento box and Ansible 2.3.1 / Vagrant 1.9.7? I haven't been able to provision staging...

```
TASK [build-securedrop-app-code-deb-pkg : Copy install_files/securedrop-app-code dir to build path.] ***
fatal: [build]: FAILED! => {"changed": false, "cmd": "/usr/bin/rsync --delay-updates -F --compress --delete-after --archive --rsh=/usr/bin/ssh -S none -i /home/kevin/.vagrant.d/insecure_private_key -o Port=2200 -o StrictHostKeyChecking=no --rsync-path='sudo rsync' --out-format=<<CHANGED>>%i %n%L /home/kevin/dev/securedrop/install_files/securedrop-app-code/ vagrant@127.0.0.1:/tmp/securedrop-app-code-0.3.12-amd64", "failed": true, "msg": "bash: sudo rsync: command not found\nrsync: connection unexpectedly closed (0 bytes received so far) [sender]\nrsync error: remote command not found (code 127) at io.c(235) [sender=3.1.2]\n", "rc": 127}
```

**Comment by redshiftzero** [comment]
What version of rsync do you have? If it's 2.x you might need to [install a more recent version](https://docs.securedrop.org/en/latest/development/getting_started.html#mac-os-x). I'm using version 3.1.2, which I installed using brew.

**Comment by conorsch** [comment]
If @ageis can confirm it's an outdated rsync version causing the problem, we should probably add a quick check to the build playbook that will fail with an informative error message. Also, note that the [dev docs]() stipulate running `pip install -r securedrop/requirements/develop-requirements.txt`, which pegs Ansible at 2.2.1—which may change, but it's the "correct" version for current development.

**Comment by ageis** [comment]
I don't think that's the problem, I figured it was that privilege issue mentioned in the related commit; I'm running rsync 3.1.2 which is the absolute latest on my host machine. The VM has 3.1.0. It would seem to be because of something that's changed in Ansible, because 2.2.1 works for me. Thought I had tried downgrading to that before but I guess not.

It might be this bug which has been filed against Ansible 2.3.x: https://github.com/ansible/ansible/issues/23575 Hopefully the bugfix will land in the next point release.

Thanks!
